src/vivarium/framework/lifecycle.py

```diff
@@ -551,8 +560,11 @@ def add_handlers(self, state_name: str, handlers: List[Callable]) -> None:
         s.add_handlers(handlers)
 
     def add_constraint(
-        self, method: Callable, allow_during: List[str] = [], restrict_during: List[str] = []
-    ):
+        self,
+        method: Callable[[Any], Any],
+        allow_during: list[str] = [],
```

Comment:
Default arguments should be immutable (and lists aren't). We should have a default value of `()` or `None`